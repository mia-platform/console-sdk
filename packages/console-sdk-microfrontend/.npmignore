.github
.storybook
.yarn
.vscode
*.tgz
**/coverage
.editorconfig
.eslintignore
.eslintrc.cjs
.gitattributes
.gitignore
.nvmrc
.yarnrc.yml
.changeset
CONTRIBUTING.md
index.html
jest.config.js
tsconfig.json
tsconfig.node.json
vite.config.ts
yarn.lock
stats.html
storybook-static
__tests__
node_modules
src/*