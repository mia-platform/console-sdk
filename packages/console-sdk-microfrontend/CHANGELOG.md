# @mia-platform/console-sdk-microfrontend

## 0.0.7

### Patch Changes

- Updated dependencies []:
  - @mia-platform/console-types@0.18.4

## 0.0.6

### Patch Changes

- Updated dependencies []:
  - @mia-platform/console-types@0.18.3

## 0.0.5

### Patch Changes

- Updated dependencies []:
  - @mia-platform/console-types@0.18.2

## 0.0.4

### Patch Changes

- Update dependencies

## 0.0.3

### Patch Changes

- Update dependencies

## 0.0.2

### Added

- ðŸš€ New getMicrofrontendNode to console-sdk

## 0.0.1

### Added

- ðŸš€ Setup repo
