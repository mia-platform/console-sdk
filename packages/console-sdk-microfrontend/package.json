{
  "name": "@mia-platform/console-sdk-microfrontend",
  "license": "Apache-2.0",
  "version": "0.0.9",
  "main": "./build/index.js",
  "typings": "./build/index.d.ts",
  "module": "./build/index.js",
  "scripts": {
    "preinstall": "npx only-allow pnpm",
    "build": "rm -rf ./build && tsc --build",
    "lint": "eslint . --ext .ts",
    "test": "jest --coverage",
    "unit": "jest"
  },
  "dependencies": {
    "@mia-platform/console-types": "workspace:^",
    "rxjs": "^7.8.1",
    "ramda": "^0.29.1"
  },
  "devDependencies": {
    "@typescript-eslint/eslint-plugin": "^7.2.0",
    "@typescript-eslint/parser": "^7.1.0",
    "@types/ramda": "^0.29.11",
    "typescript": "^5.3.3",
    "vite-plugin-qiankun": "1.0.15",
    "@types/jest": "^29.5.12",
    "@types/jsdom": "21.1.6",
    "jest": "^29.7.0",
    "ts-jest": "^29.1.2"
  },
  "engines": {
    "node": ">=v18.18.2"
  },
  "homepage": "https://github.com/mia-platform/console-sdk-microfrontend#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/mia-platform/console-sdk-microfrontend.git"
  },
  "bugs": {
    "url": "https://github.com/mia-platform/console-sdk-microfrontend/issues"
  },
  "jest": {
    "preset": "ts-jest",
    "modulePathIgnorePatterns": [
      "<rootDir>/build"
    ],
    "testMatch": [
      "**/__tests__/**/*.+(ts)",
      "**/?(*.)+(spec|test).+(ts)"
    ]
  }
}
