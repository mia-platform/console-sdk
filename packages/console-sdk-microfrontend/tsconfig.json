{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "./build",
    "rootDir": "src/"
  },
  "exclude": [
    "./node_modules",
    "./coverage",
    "./build",
    "./**/*.test.ts",
    "tsconfig.json"
  ],
  "include": [
    "./**/*.ts",
    "./**/*.json"
  ],
}
