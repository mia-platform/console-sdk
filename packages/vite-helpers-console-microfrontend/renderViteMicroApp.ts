import { ConsoleSDK, IConsoleSDK } from '@mia-platform/microfrontend-sdk'
import { renderWithQiankun, qiankunWindow, QiankunProps } from 'vite-plugin-qiankun/dist/helper'

export type ConsoleLifecycleFunction = (
  isConnectedToConsole: boolean,
  consoleSDK: IConsoleSDK,
) => void

export interface IViteParams {
  mount: ConsoleLifecycleFunction
  unmount: ConsoleLifecycleFunction
  bootstrap?: () => void
  update?: () => void
}

export function getSDK (props: QiankunProps): {
  consoleSDK: IConsoleSDK
  isConnectedToConsole: boolean
} {
  const consoleSDK = new ConsoleSDK(props)
  const isConnectedToConsole = Boolean(qiankunWindow.__POWERED_BY_QIANKUN__)

  return {
    consoleSDK,
    isConnectedToConsole
  }
}

export const decorateLifecycleFunction = (lifecycleFunction: IViteParams[keyof IViteParams]) => {
  return (props: QiankunProps): void => {
    const { isConnectedToConsole, consoleSDK } = getSDK(props)
    lifecycleFunction && lifecycleFunction(isConnectedToConsole, consoleSDK)
  }
}

export default function renderViteMicroApp ({ mount, unmount }: IViteParams) {
  renderWithQiankun({
    mount: decorateLifecycleFunction(mount),
    unmount: decorateLifecycleFunction(unmount),
    bootstrap () {},
    update () {}
  })
}
