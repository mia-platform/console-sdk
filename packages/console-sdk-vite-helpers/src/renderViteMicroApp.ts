/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ConsoleSDK, IConsoleProps, IMicrofronendIntegrator } from '@mia-platform/console-sdk-microfrontend'
import { QiankunProps, qiankunWindow, renderWithQiankun } from 'vite-plugin-qiankun/dist/helper'

export type ConsoleLifecycleFunction = (
  isConnectedToConsole: boolean,
  consoleSDK: IMicrofronendIntegrator,
) => void

export type IViteParams = {
  mount: ConsoleLifecycleFunction
  unmount: ConsoleLifecycleFunction
  bootstrap?: () => void
  update?: () => void
}

export function getSDK(props: QiankunProps): {
  consoleSDK: IMicrofronendIntegrator
  isConnectedToConsole: boolean
} {
  const consoleSDK = new ConsoleSDK(props as IConsoleProps)
  const isConnectedToConsole = Boolean(qiankunWindow.__POWERED_BY_QIANKUN__)

  return {
    consoleSDK,
    isConnectedToConsole,
  }
}

export const decorateLifecycleFunction = (lifecycleFunction: IViteParams[keyof IViteParams]) => {
  return (props: QiankunProps): void => {
    const { isConnectedToConsole, consoleSDK } = getSDK(props)
    if (lifecycleFunction) {
      lifecycleFunction(isConnectedToConsole, consoleSDK)
    }
  }
}

export default function renderViteMicroApp({ mount, unmount }: IViteParams): void {
  renderWithQiankun({
    mount: decorateLifecycleFunction(mount),
    unmount: decorateLifecycleFunction(unmount),
    bootstrap: () => undefined,
    update: () => undefined,
  })
}
