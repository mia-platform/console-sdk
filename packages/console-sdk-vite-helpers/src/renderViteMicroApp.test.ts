/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ConsoleSDK } from '@mia-platform/console-sdk-microfrontend'
import { JSDOM } from 'jsdom'
import { QiankunProps } from 'vite-plugin-qiankun/dist/helper'

import { IViteParams, decorateLifecycleFunction, getSDK } from './renderViteMicroApp'

const dom = new JSDOM()
const { document } = dom.window

const quiankunContainer = document.createElement('div')
quiankunContainer.id = '__qiankun_microapp_wrapper_for_microfrontend__'

const qiankunPropsMock: QiankunProps = {
  name: 'testMicrofrontend',
  container: quiankunContainer,
  // some console injected props
  console: {},
  eventListener: jest.fn(),
  resourceAPI: { writeConfig: jest.fn() },
}

const viteParamsMock: IViteParams = {
  mount: jest.fn(),
  unmount: jest.fn(),
}

describe('Vite Micro App Rendering', () => {
  it('should get lifecycle function params', () => {
    const { isConnectedToConsole, consoleSDK } = getSDK(qiankunPropsMock)
    expect(isConnectedToConsole).toBe(false)
    expect(consoleSDK).toBeInstanceOf(ConsoleSDK)
  })

  it('should mount the micro app', () => {
    decorateLifecycleFunction(viteParamsMock.mount)(qiankunPropsMock)
    expect(viteParamsMock.mount).toHaveBeenCalledWith(false, expect.any(ConsoleSDK))
  })

  it('should mount the micro app on the missing props', () => {
    decorateLifecycleFunction(viteParamsMock.mount)()
    decorateLifecycleFunction(viteParamsMock.unmount)()

    expect(viteParamsMock.mount).toHaveBeenCalledWith(false, undefined)
    expect(viteParamsMock.unmount).toHaveBeenCalledWith(false, undefined)
  })

  it('should mount the micro app on empty props', () => {
    decorateLifecycleFunction(viteParamsMock.mount)({})
    decorateLifecycleFunction(viteParamsMock.unmount)({})

    expect(viteParamsMock.mount).toHaveBeenCalledWith(false, undefined)
    expect(viteParamsMock.unmount).toHaveBeenCalledWith(false, undefined)
  })

  it('should unmount the micro app', () => {
    decorateLifecycleFunction(viteParamsMock.unmount)(qiankunPropsMock)
    expect(viteParamsMock.unmount).toHaveBeenCalledWith(false, expect.any(ConsoleSDK))
  })
})
