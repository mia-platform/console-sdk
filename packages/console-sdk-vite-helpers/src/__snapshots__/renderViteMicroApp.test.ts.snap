// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vite Micro App Rendering should mount the micro app on the missing props 1`] = `
[MockFunction] {
  "calls": [
    [
      false,
      MicrofrontendIntegrator {
        "configObservable": {
          "_version": "0.2.32",
          "collections": undefined,
          "configMaps": undefined,
          "endpoints": undefined,
          "forceConfigUpdateChecksum": undefined,
          "microfrontendPluginConfig": undefined,
          "selectedEnvironment": undefined,
          "selectedProject": undefined,
          "services": undefined,
          "unsecretedVariables": undefined,
        },
        "consoleWriteConfig": [MockFunction],
        "container": <div
          id="__qiankun_microapp_wrapper_for_microfrontend__"
        />,
        "contexts": {
          "featureTogglesProxyContext": undefined,
          "hotkeysContext": undefined,
        },
        "events": Subject {
          "closed": false,
          "currentObservers": null,
          "hasError": false,
          "isStopped": false,
          "observers": [
            SafeSubscriber {
              "_finalizers": [
                Subscription {
                  "_finalizers": null,
                  "_parentage": [Circular],
                  "closed": false,
                  "initialTeardown": [Function],
                },
              ],
              "_parentage": null,
              "closed": false,
              "destination": ConsumerObserver {
                "partialObserver": {
                  "complete": undefined,
                  "error": undefined,
                  "next": [MockFunction],
                },
              },
              "initialTeardown": undefined,
              "isStopped": false,
            },
          ],
          "thrownError": null,
        },
        "name": "testMicrofrontend",
      },
    ],
    [
      false,
      MicrofrontendIntegrator {
        "configObservable": {
          "_version": "0.2.32",
          "collections": undefined,
          "configMaps": undefined,
          "endpoints": undefined,
          "forceConfigUpdateChecksum": undefined,
          "microfrontendPluginConfig": undefined,
          "selectedEnvironment": undefined,
          "selectedProject": undefined,
          "services": undefined,
          "unsecretedVariables": undefined,
        },
        "consoleWriteConfig": undefined,
        "container": undefined,
        "contexts": {
          "featureTogglesProxyContext": undefined,
          "hotkeysContext": undefined,
        },
        "events": Subject {
          "closed": false,
          "currentObservers": null,
          "hasError": false,
          "isStopped": false,
          "observers": [
            SafeSubscriber {
              "_finalizers": [
                Subscription {
                  "_finalizers": null,
                  "_parentage": [Circular],
                  "closed": false,
                  "initialTeardown": [Function],
                },
              ],
              "_parentage": null,
              "closed": false,
              "destination": ConsumerObserver {
                "partialObserver": {
                  "complete": undefined,
                  "error": undefined,
                  "next": undefined,
                },
              },
              "initialTeardown": undefined,
              "isStopped": false,
            },
          ],
          "thrownError": null,
        },
        "name": undefined,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Vite Micro App Rendering should mount the micro app on the missing props 2`] = `
[MockFunction] {
  "calls": [
    [
      false,
      MicrofrontendIntegrator {
        "configObservable": {
          "_version": "0.2.32",
          "collections": undefined,
          "configMaps": undefined,
          "endpoints": undefined,
          "forceConfigUpdateChecksum": undefined,
          "microfrontendPluginConfig": undefined,
          "selectedEnvironment": undefined,
          "selectedProject": undefined,
          "services": undefined,
          "unsecretedVariables": undefined,
        },
        "consoleWriteConfig": undefined,
        "container": undefined,
        "contexts": {
          "featureTogglesProxyContext": undefined,
          "hotkeysContext": undefined,
        },
        "events": Subject {
          "closed": false,
          "currentObservers": null,
          "hasError": false,
          "isStopped": false,
          "observers": [
            SafeSubscriber {
              "_finalizers": [
                Subscription {
                  "_finalizers": null,
                  "_parentage": [Circular],
                  "closed": false,
                  "initialTeardown": [Function],
                },
              ],
              "_parentage": null,
              "closed": false,
              "destination": ConsumerObserver {
                "partialObserver": {
                  "complete": undefined,
                  "error": undefined,
                  "next": undefined,
                },
              },
              "initialTeardown": undefined,
              "isStopped": false,
            },
          ],
          "thrownError": null,
        },
        "name": undefined,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
